### DISCUSSION TITLE
Calling a store method from another store method does not update the state

### DISCUSSION THREAD

**Discussion by vcardins** [pmndrs/zustand#2265] [General]
I'm trying to call a store internal method to update some state values but it doesn't and the component is not re-rendered. On the other hand, if I update the state directly (I'm using immer) it works. What am I doing wrong?

![image](https://github.com/pmndrs/zustand/assets/2383329/86ead9fd-3b11-4574-a6a2-773f7cb21523)

**Comment by dai-shi** [comment]
I think mixing normal update and immer update causes the problem. It should work without immer middleware.

**Reply by vcardins** [reply]
I removed `immer` and it still didn't work. Does it have anything to do with how the `setShowVisualizationConfigurationPanel` method is written? Is there any recommendation for cases like this?

**Reply by vcardins** [reply]
`immer` is crucial for my project once I'm migrating some cold that "mutate" the state. Any idea on how I can achieve it still using `immer`?

**Reply by vcardins** [reply]
I found https://github.com/pmndrs/zustand/discussions/307, which is having the same issue.

**Reply by charkour** [reply]
What if you replace `get().setShow(false)` to `state.setShow(false)`?

**Reply by vcardins** [reply]
Same thing

**Reply by charkour** [reply]
Oh, looking through the code, because you have `state` in scope of the callback, `get().setShow()` is likely modifying the current `state` object and not the local `state` value in your callback you're returning which is why the second version works but not the first. 

Try this: 

```ts
resetSearch: () => set((state) => { 
  state.setShow(false)
  return { items: [] }
})


// Or more simply
resetSearch: () => set({ items: [], isShowing: false })
```
